server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimGroupProvisioning.java

```diff
@@ -220,9 +229,24 @@ public ScimGroup retrieve(String id, final String zoneId) throws ScimResourceNot
         }
     }
 
+    @SuppressWarnings("java:S1874")
+    private Set<String> getAllowedUserGroups(String zoneId) {
+        Set<String> zoneAllowedGroups = null; // default: all groups allowed
+        try {
+            IdentityZone currentZone = IdentityZoneHolder.get();
+            zoneAllowedGroups = (currentZone.getId().equals(zoneId)) ?
+                currentZone.getConfig().getUserConfig().resultingAllowedGroups() :
+                jdbcIdentityZoneProvisioning.retrieve(zoneId).getConfig().getUserConfig().resultingAllowedGroups();
```

Comment:
Hi @bruce-ricard ok, so vote to stay here with the logic . We may can reactor readability, but the current zone is in the local thread already, no DB lookup needed, but other zones need to be retrieved from DB. if we simply retrieve from DB, then we have always a DB lookup

Reply:
OK so it's a performance optimization. It's what I was thinking.

I like to try to make code self explanatory, whenever possible. For example here, I don't want future readers of the code to ask themselves this question I asked, until the end of times. 
Typically, I would have created a new method for this operation, and tried to pass the performance optimization explanation to the reader through good variable and method names.

Another different argument would be to say: this method is only called at group update and creation time. This is quite rare. I don't think that we should care about a single DB lookup made at group write time.

Something else: we do the test `currentZone.getId().equals(zoneId)`, and on the next line we get the current zone config. Isn't it possible that between these 2 instructions, another thread changes the current zone, and we end up looking at allowed groups from another zone?

Reply:
Regarding the last paragraph: the same may happen if you use an if-statement instead of a conditional. And the time elapsed between the two instructions is really tiny :-)